### PR TITLE
feat: client.addMetadataFilter(fn)

### PR DESCRIPTION
**Pull request by trentm**
A facility to custom filter the "metadata" object sent to APM server at
the start of every intake request.

This refactors setting of `_encodedMetadata` to happen in only one
place (`_resetEncodedMetadata`), so that it can apply the metadata
filters in only on place.

Refs: https://github.com/elastic/apm-agent-nodejs/issues/1916

### CODE REVIEW

**Comment by trentm** [comment]
The matching agent PR is https://github.com/elastic/apm-agent-nodejs/pull/2045

File: README.md [hunk 1]
```diff
@@ -260,6 +260,37 @@ configuration options can be updated except:
...
+
+```
+[2021-04-14T22:28:35.419Z] ERROR (elastic-apm-node): APM Server transport error (400): Unexpected APM Server response
+APM Server accepted 0 events in the last request
+Error: validation error: 'metadata' required
+  Document: {"metadata":null}
+```
+
+See the [APM Agent `addMetadataFilter` documentation](https://www.elastic.co/guide/en/apm/agent/nodejs/current/agent-api.html#apm-add-metadata-filter)
+for further details.
```
Comment (astorm) [inline README.md:291]: saying it out loud -- [this](https://github.com/elastic/apm-agent-nodejs/pull/2045/files#diff-23427ac409c61f7de0225443c0c7c77aca90b732e73957bad73edb357c30addeR163) will need to land before this link works.

File: index.js [hunk 2]
```diff
@@ -231,20 +237,43 @@ Client.prototype.config = function (opts) {
   // fixes bug where cached/memoized _encodedMetadata wouldn't be
   // updated when client was reconfigured
   if (this._encodedMetadata) {
-    this.updateEncodedMetadata()
+    this._resetEncodedMetadata()
```
Comment (astorm) [inline index.js:240]: Is it worth mentioning in the docs that filters will be applied _whenever_ the agent is reconfigured? (not just on startup)?

**Review by astorm — APPROVED** [review]
TL;DR; -- does what it says on the tin -- one question below about some possible docs additions, but approving.  Also, here's what I see i this PR

I see we've moved our logic for encoding the metadata from the `_fetchAndEncodeMetadata` method to a new `_resetEncodedMetadata` method.  The `_fetchAndEncodeMetadata` method is still responsible for initiating the _cloud_ metadata request, but all other work has been passed on to the `_resetEncodedMetadata` method. We've also added a new `_cloudMetadata` property in order to coordinate between `_fetchAndEncodeMetadata` and `_resetEncodedMetadata`.  The `_fetchAndEncodeMetadata` method sets this variable, and `_resetEncodedMetadata` is responsible for reading it. 

The new `_resetEncodedMetadata` method does a deep clone of the metadata stored at `this._conf.metadata` and, if present, the cloud metadata stored at `this._cloudMetadata`.   This deeply cloned object is passed through any callback/filter functions registered with the new `Filter/`object-filter-sequence` object.  The final filtered object is set as the `_encodedMetadata` object.

Speaking of filters, we've also added a public `addMetadataFilter` function that allows us to register a callback function with the `Filter/`object-filter-sequence` instance used by the client. 

Finally, the client invokes the `_resetEncodedMetadata` method in three places

1. After the initial fetch of the cloud metadata.
2. Whenever the client is reconfigured. 
3. When a filter is added IF `_encodedMetadata` has been set.

File: README.md (hunk 1 above)
Comment (trentm) [reply README.md:291]: Yup.

File: index.js (hunk 2 above)
Comment (trentm) [reply index.js:240]: Yup, that is worth it. Added in commit 9a12928
